### ISSUE TITLE
Potential dangling pointers from caching behavior

### ISSUE THREAD

**Issue by jdaignault** [DoubleDeez/MDFastBinding#37]
Commit https://github.com/DoubleDeez/MDFastBinding/commit/3162484e645ae6a0048c06c7c4e68446543d5b3f added some logic to cache various pointers and minimize redundant logic, however there are multiple cases where the cached pointers can dangle and result in undefined behavior. Here are two scenarios where this has happened:

**Steps to reproduce:**

Scenario 1:
1. Follow the repro steps outlined in issue #22 
2. Observe that after performing step 7, the application crashes (this may or may not happen since there is undefined behavior)
-- In our case we crashed within `FMDFastBindingPropertySetter_Objects::CanSetProperty`

**Workaround**: Comment out the following line within `FMDFastBindingFunctionWrapper::GetParams()`:

```
if (!LastFrameFunctionUpdated.IsSet() || LastFrameFunctionUpdated.GetValue() != GFrameCounter)
```

Scenario 2:
1. Enable diffing for binding graphs
2. Open a widget blueprint asset that contains existing bindings
3. Compile the blueprint
4. Observe that it crashes (this may or may not happen since there is undefined behavior)
-- In our case we crashed within `UMDFastBindingGraphNode::AllocateDefaultPins`

**Workaround**: Comment out the following line within `FMDFastBindingFieldPath::GetFieldPath()`:

```
if (!LastFrameUpdatedPath.IsSet() || LastFrameUpdatedPath.GetValue() != GFrameCounter)
```

**Comment by DoubleDeez** [comment]
> As a stopgap, is it safe to disable/remove the `LastFrame` caching guards for Editor platforms? Other than bringing back some redundant processing, would it introduce any other regressions or issues?

My memory is a bit hazy on it, but IIRC there were dangling pointers/memory stomps/_something bad_ caused by _not_ caching for the frame as they would get stomped by subsequent calls in the same frame.

I'll have to investigate further to figure out why I needed to add it.

**Comment by jdaignault** [comment]
Yeah, it looks like I eventually hit that kind of behavior too with these guards omitted :(

Do you think it seems sensible to instead rely on more `TWeakFieldPtr` usage across the board, with possible raw pointer optimizations for non-editor targets? At the very least it would convert the issues from undefined behavior into runtime errors, which are a lot easier to deal with.

**Comment by DoubleDeez** [comment]
Are you seeing the issue without any bindings executing? (Not running the game and with design-time bindings disabled).

I thought the cause was with the allocated memory, not the FProperty* ptrs.

**Comment by jdaignault** [comment]
The issue is with the FProperty pointers yeah, these crashes occur at design time without design-time bindings enabled.

**Comment by jdaignault** [comment]
Alright, I think I've tracked down a bunch of issues:

* Converting more raw pointers to use `TWeakFieldPtr` indeed exposed some issues where field owners were no longer valid, which often resulted in binding values getting orphaned
* After enabling fastbing diffs, it appears that the implementation of `UMDFastBindingWidgetBlueprintExtension::GetAllGraphs` turned out to be quite fragile, notably during the line `GraphObj->SetBinding(BindingInstance);`. GetAllGraphs can be called quite frequently during a blueprint compilation context, also resulting in bindings getting orphaned
* There was an issue present in `UMDFastBindingDestination_Function::SetupBindingItems` during a blueprint compilation context.  -- It would call `OrphanBindingItem` before `BindingItems.RemoveAt(i);`
-- `UMDFastBindingInstance::AddOrphan` would invoke `DuplicateObject` on a value
-- This triggered a PostLoad, and apparently this allowed some code to cycle back to another round of `UMDFastBindingDestination_Function::SetupBindingItems` on the same object
-- This ended up calling `BindingItems.RemoveAt(i);` on the same index, then eventually it would go back up the stack and crash trying to resolve the original call to `BindingItems.RemoveAt(i);` (as it was removed during the PostLoad context)

I have some fixes locally that are starting to shape up, so I might try to stage some PRs soon.
